File: host/protocol.ts
export class ProtocolError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "ProtocolError";
  }
}

export interface Hello {
  type: "hello";
  mac: string;
  chip: string;
}

export interface Ready {
  type: "ready";
  role: "TX" | "RX";
  mac: string;
  channel: number;
  detail: string;
}

export interface DeviceError {
  type: "error";
  message: string;
}

export interface CsiFrame {
  type: "csi";
  receiver: string;
  sequence: number;
  timestampUs: number;
  rssi: number;
  noiseFloor: number;
  channel: number;
  dropped: number;
  samples: number[];
}

export type Message = Hello | Ready | DeviceError | CsiFrame;

export function normalizeMac(value: string): string {
  const compact = value.replaceAll(":", "").replaceAll("-", "").toLowerCase();
  if (compact.length !== 12) {
    throw new ProtocolError("MAC address must contain 12 hexadecimal digits");
  }
  if (!/^[0-9a-f]{12}$/.test(compact)) {
    throw new ProtocolError("MAC address is not hexadecimal");
  }
  return compact.match(/.{2}/g)!.join(":");
}

function parseInteger(value: string): number {
  if (!/^[+-]?\d+$/.test(value)) {
    throw new Error("invalid integer");
  }
  const parsed = Number(value);
  if (!Number.isSafeInteger(parsed)) {
    throw new Error("integer is outside the supported range");
  }
  return parsed;
}

/** Parse a firmware record, ignoring ESP-IDF logs and empty lines. */
export function parseLine(line: string): Message | null {
  const stripped = line.trim();
  if (!stripped || !stripped.startsWith("RADAR,")) {
    return null;
  }

  const parts = stripped.split(",");
  if (parts.length < 2) {
    throw new ProtocolError("record has no type");
  }

  const messageType = parts[1];
  if (messageType === "HELLO") {
    if (parts.length !== 4) {
      throw new ProtocolError("HELLO requires MAC and chip");
    }
    return { type: "hello", mac: normalizeMac(parts[2]), chip: parts[3] };
  }
  if (messageType === "READY") {
    if (parts.length !== 6 || (parts[2] !== "TX" && parts[2] !== "RX")) {
      throw new ProtocolError("READY has an invalid role");
    }
    let channel: number;
    try {
      channel = parseInteger(parts[4]);
    } catch {
      throw new ProtocolError("READY contains an invalid channel");
    }
    if (channel < 1 || channel > 13) {
      throw new ProtocolError("READY channel is outside the supported 2.4 GHz range");
    }
    return {
      type: "ready",
      role: parts[2],
      mac: normalizeMac(parts[3]),
      channel,
      detail: parts[5],
    };
  }
  if (messageType === "ERROR") {
    if (parts.length !== 3) {
      throw new ProtocolError("ERROR requires one message");
    }
    return { type: "error", message: parts[2] };
  }
  if (messageType !== "CSI") {
    return null;
  }
  if (parts.length !== 11) {
    throw new ProtocolError("CSI requires nine metadata fields and a payload");
  }

  let receiver: string;
  let sequence: number;
  let timestampUs: number;
  let rssi: number;
  let noiseFloor: number;
  let channel: number;
  let dropped: number;
  let declaredLength: number;
  let rawSamples: number[];
  try {
    receiver = normalizeMac(parts[2]);
    sequence = parseInteger(parts[3]);
    timestampUs = parseInteger(parts[4]);
    rssi = parseInteger(parts[5]);
    noiseFloor = parseInteger(parts[6]);
    channel = parseInteger(parts[7]);
    dropped = parseInteger(parts[8]);
    declaredLength = parseInteger(parts[9]);
    if (parts[10].length % 2 !== 0 || !/^[0-9a-f]*$/i.test(parts[10])) {
      throw new Error("invalid hexadecimal data");
    }
    rawSamples = parts[10].match(/.{2}/g)?.map((value) => Number.parseInt(value, 16)) ?? [];
  } catch {
    throw new ProtocolError("CSI contains invalid numeric or hexadecimal data");
  }

  if (declaredLength !== rawSamples.length) {
    throw new ProtocolError(
      `CSI payload length is ${rawSamples.length}, expected ${declaredLength}`,
    );
  }
  if (declaredLength === 0 || declaredLength % 2 !== 0) {
    throw new ProtocolError("CSI payload must contain complete I/Q pairs");
  }
  if (channel < 1 || channel > 13) {
    throw new ProtocolError("CSI channel is outside the supported 2.4 GHz range");
  }
  if (dropped < 0) {
    throw new ProtocolError("CSI dropped count cannot be negative");
  }

  const samples = rawSamples.map((value) => (value < 128 ? value : value - 256));
  return {
    type: "csi",
    receiver,
    sequence,
    timestampUs,
    rssi,
    noiseFloor,
    channel,
    dropped,
    samples,
  };
}

File: host/protocol.py
"""Line protocol shared with the ESP32 firmware."""

from __future__ import annotations

from dataclasses import dataclass


class ProtocolError(ValueError):
    """Raised when a RADAR line is present but malformed."""


@dataclass(frozen=True)
class Hello:
    mac: str
    chip: str


@dataclass(frozen=True)
class Ready:
    role: str
    fields: tuple[str, ...]


@dataclass(frozen=True)
class DeviceError:
    message: str


@dataclass(frozen=True)
class CsiFrame:
    receiver: str
    sequence: int
    timestamp_us: int
    rssi: int
    noise_floor: int
    channel: int
    dropped: int
    samples: tuple[int, ...]


Message = Hello | Ready | DeviceError | CsiFrame


def normalize_mac(value: str) -> str:
    compact = value.replace(":", "").replace("-", "").lower()
    if len(compact) != 12:
        raise ProtocolError("MAC address must contain 12 hexadecimal digits")
    try:
        bytes.fromhex(compact)
    except ValueError as error:
        raise ProtocolError("MAC address is not hexadecimal") from error
    return ":".join(compact[index : index + 2] for index in range(0, 12, 2))


def parse_line(line: str) -> Message | None:
    """Parse a firmware record, ignoring ESP-IDF logs and empty lines."""

    stripped = line.strip()
    if not stripped or not stripped.startswith("RADAR,"):
        return None

    parts = stripped.split(",")
    if len(parts) < 2:
        raise ProtocolError("record has no type")

    message_type = parts[1]
    if message_type == "HELLO":
        if len(parts) != 4:
            raise ProtocolError("HELLO requires MAC and chip")
        return Hello(normalize_mac(parts[2]), parts[3])
    if message_type == "READY":
        if len(parts) < 4 or parts[2] not in {"TX", "RX"}:
            raise ProtocolError("READY has an invalid role")
        return Ready(parts[2], tuple(parts[3:]))
    if message_type == "ERROR":
        if len(parts) != 3:
            raise ProtocolError("ERROR requires one message")
        return DeviceError(parts[2])
    if message_type != "CSI":
        return None
    if len(parts) != 11:
        raise ProtocolError("CSI requires nine metadata fields and a payload")

    try:
        receiver = normalize_mac(parts[2])
        sequence = int(parts[3])
        timestamp_us = int(parts[4])
        rssi = int(parts[5])
        noise_floor = int(parts[6])
        channel = int(parts[7])
        dropped = int(parts[8])
        declared_length = int(parts[9])
        raw_samples = bytes.fromhex(parts[10])
    except ValueError as error:
        raise ProtocolError("CSI contains invalid numeric or hexadecimal data") from error

    if declared_length != len(raw_samples):
        raise ProtocolError(
            f"CSI payload length is {len(raw_samples)}, expected {declared_length}"
        )
    if declared_length == 0 or declared_length % 2 != 0:
        raise ProtocolError("CSI payload must contain complete I/Q pairs")
    if not 1 <= channel <= 13:
        raise ProtocolError("CSI channel is outside the supported 2.4 GHz range")
    if dropped < 0:
        raise ProtocolError("CSI dropped count cannot be negative")

    samples = tuple(value if value < 128 else value - 256 for value in raw_samples)
    return CsiFrame(
        receiver=receiver,
        sequence=sequence,
        timestamp_us=timestamp_us,
        rssi=rssi,
        noise_floor=noise_floor,
        channel=channel,
        dropped=dropped,
        samples=samples,
    )
